lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,730 @@
+
+  if (auto *fri = dyn_cast<FunctionRefInst>(value))
+    return SymbolicValue::getFunction(fri->getReferencedFunction());
+
+  // If we have a reference to a metatype, constant fold any substitutable
+  // types.
+  if (auto *mti = dyn_cast<MetatypeInst>(value)) {
+    auto metatype = mti->getType().castTo<MetatypeType>();
+    auto type = simplifyType(metatype->getInstanceType())->getCanonicalType();
+    return SymbolicValue::getMetatype(type);
+  }
+
+  if (auto *tei = dyn_cast<TupleExtractInst>(value)) {
+    auto val = getConstantValue(tei->getOperand());
+    if (!val.isConstant())
+      return val;
+    return val.getAggregateValue()[tei->getFieldNo()];
+  }
+
+  // If this is a struct extract from a fragile type, then we can return the
+  // element being extracted.
+  if (auto *sei = dyn_cast<StructExtractInst>(value)) {
+    auto aggValue = sei->getOperand();
+    auto val = getConstantValue(aggValue);
+    if (val.isConstant()) {
+      assert(val.getKind() == SymbolicValue::Aggregate);
+      return val.getAggregateValue()[sei->getFieldNo()];
+    }
+    // Not a const.
+    return val;
+  }
+
+  // TODO: If this is a single element struct, we can avoid creating an
+  // aggregate to reduce # allocations.  This is extra silly in the case of zero
+  // element tuples.
+  if (isa<StructInst>(value) || isa<TupleInst>(value)) {
+    auto inst = cast<SingleValueInstruction>(value);
+    SmallVector<SymbolicValue, 4> elts;
+
+    for (unsigned i = 0, e = inst->getNumOperands(); i != e; ++i) {
+      auto val = getConstantValue(inst->getOperand(i));
+      if (!val.isConstant())
+        return val;
+      elts.push_back(val);
+    }
+
+    return SymbolicValue::getAggregate(elts, evaluator.getAllocator());
+  }
+
+  if (auto *builtin = dyn_cast<BuiltinInst>(value))
+    return computeConstantValueBuiltin(builtin);
+
+  if (auto *apply = dyn_cast<ApplyInst>(value)) {
+    auto callResult = computeCallResult(apply);
+
+    // If this failed, return the error code.
+    if (callResult.hasValue())
+      return callResult.getValue();
+
+    assert(calculatedValues.count(apply));
+    return calculatedValues[apply];
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Unknown simple: " << *value << "\n");
+
+  // Otherwise, we don't know how to handle this.
+  return evaluator.getUnknown(value, UnknownReason::Default);
+}
+
+SymbolicValue
+ConstExprFunctionState::computeConstantValueBuiltin(BuiltinInst *inst) {
+  const BuiltinInfo &builtin = inst->getBuiltinInfo();
+
+  // Handle various cases in groups.
+  auto unknownResult = [&]() -> SymbolicValue {
+    return evaluator.getUnknown(SILValue(inst), UnknownReason::Default);
+  };
+
+  // Unary operations.
+  if (inst->getNumOperands() == 1) {
+    auto operand = getConstantValue(inst->getOperand(0));
+    // TODO: Could add a "value used here" sort of diagnostic.
+    if (!operand.isConstant())
+      return operand;
+
+    // TODO: SUCheckedConversion/USCheckedConversion
```

Comment:
@marcrasi  Is this SUCheckerConversion is what is being implemented in the following lines? Is it a TODO still?

Reply:
It's still a TODO. The following lines implement "xxxCheckedTrunc", not "xxxCheckedConversion".

"xxxCheckedConversion" seems to be a builtin that exists but that the stdlib never calls.